### DISCUSSION TITLE
Auto-incrementing column without primary key

### DISCUSSION THREAD

**Discussion by grimmdude** [laravel/framework#45775] [Ideas]
I have a need to create an auto-incrementing column which is not to be used as the primary key of the table.  There are methods like `$table->bigIncrements('id')` which create auto-increment columns, but they also assume the column as the primary key.

It seems like it would be ideal to separate the primary key concerns from these methods so the developer can make that choice.  For instance, by using the `->primary()` method:

```php
// Non-primary
$table->bigIncrements('column');

// Primary
$table->bigIncrements('column')->primary();
```

Though, I guess at this point it would be a breaking change to do that.  Perhaps the solution would be to create new methods specific to non-primary key use?

```php
$table->bigIncrementsNonPrimary('column');
```

**Comment by henzeb** [comment]
You mean:
```php
$table->bigInteger('column')->autoIncrement(); 
```
?

**Reply by grimmdude** [reply]
This method also defines the auto-increment column as the primary key.

**Reply by henzeb** [reply]
That's probably because mysql requires auto_increment columns to be part of a key.  I tried creating it with an index, but this only adds an alter statement, while indexes are perfectly able to be added to the same statement.

**Reply by grimmdude** [reply]
It is possible to create an auto-increment column in MySql which isn't the primary key though, right?

**Reply by henzeb** [reply]
Yes, but it has to be a key. So a unique index would be OK. But since laravel is executing unique indexes as a seperate statement, you need to create the table with index first, and then modify the column to add auto_increment.

**Comment by denizgolbas** [comment]
Yes there is no Schema method for making auto-increment without primary key. But i used this migration with pgsql:
```php
    public function up()
    {
        Schema::create('foos', function(Blueprint $table) {
            $table->bigInteger('id')->unsigned()->index();
            ...
        });
        DB::statement('CREATE SEQUENCE foos_id_seq START 1 owned by foos.id;');
        DB::statement('ALTER TABLE foos ALTER COLUMN id SET DEFAULT nextval(\'foos_id_seq\'::regclass)');
    }
```

**Reply by grimmdude** [reply]
Cool, yea it can be done by using `DB::statement()` but would be nice if it was possible with Schema methods.

**Comment by raghavendra89** [comment]
I agree with this. It would be good if we can create auto incrementing columns without setting them as primary key.

In our case we need an auto incrementing order number which will be used in the Invoices. But we don't want to use this order number as primary key. Because we are using a standard UUID column as primary key in all the tables.

**Comment by alazark94** [comment]
Would it affect the performance if I update it in the `static booted()` function of the model while creating the model?
Specifically like the following?

```
protected static function booted()
    {
        static::creating(function (User $user) {
            $latestUser = User::latest()->first();
            if ($latestUser) {
                $latestId = $latestUser->membership_id;
                $user->membership_id = $latestId + 1;
            } else {
                $user->membership_id = 1;
            }
        });
    }
```

**Reply by ghost** [reply]
What if you mass insert? You would not have an event for that.

**Reply by alazark94** [reply]
My current implementation uses this and it's working so far. It works with mass inserts, race conditions. You can try it by running a factory on your model and creating many users.
```
static::creating(function (User $user) {
            DB::transaction(function () use ($user) {
                DB::table('users')->lockForUpdate()->get();
                $latestIncrementalValue = User::max('membership_id');
                $newIncrementalValue = $latestIncrementalValue + 1;

                $user->membership_id = $newIncrementalValue;
            });
        });
```

**Reply by ghost** [reply]
You are locking the whole table....Anyway If you use raw inserts it will not work I guess. Raw inserts don't use the model.

**Comment by hafezdivandari** [comment]
Fixed on Laravel 11.x https://github.com/laravel/framework/pull/49925

**Reply by grimmdude** [reply]
@hafezdivandari, does this PR support creating an auto-incrementing column which isn't the primary key?  It seems like the examples all assume it's the primary key, unless I'm mistaken.

**Reply by hafezdivandari** [reply]
On MySQL, auto-increment column must be a part of primary key.

**Reply by justinrussell** [reply]
@hafezdivandari That isn't exactly correct. Per [the docs](https://dev.mysql.com/doc/refman/8.0/en/create-table.html):

> There can be only one AUTO_INCREMENT column per table, it must be indexed, and it cannot have a DEFAULT value.

If you have, for example, a table that uses UUIDs for a primary key and choose to add an auto-incrementing `display_order` field with a unique key, that's perfectly acceptable in MySQL.

**Reply by hafezdivandari** [reply]
Ok, you mean something like this?

```php
CREATE TABLE `test` (
  `uuid` char(36) NOT NULL, 
  `display_order` BIGINT UNSIGNED NOT NULL AUTO_INCREMENT,
  PRIMARY KEY (`uuid`),
  UNIQUE (`display_order`)
);
```

It's currently not supported by Laravel but I can send a PR to fix this, then you can have:

```php
Schema::create('test', function (Blueprint $table) {
    $table->uuid()->primary();
    $table->bigIncrements('display_order')->unique();
});
```

is this what you want?

**Reply by justinrussell** [reply]
If that works, I think that would be perfect.

Just wanted to clarify that MySQL allows increments that aren't a primary key.

**Reply by hafezdivandari** [reply]
Thanks for pointing that out, you may check this PR #50155

**Reply by ghost** [reply]
@hafezdivandari will this mean that the autoincrementing column will be fetched from db on each insert? Or it will behave like a default, meaning you have to create and hydrate the model to get it?

**Reply by hafezdivandari** [reply]
@marius-mcp there is `insertGetId` method for this (single query that insert and fetch the ID at once): https://laravel.com/docs/11.x/queries#auto-incrementing-ids

PS: PR #50155 has been rejected.

**Comment by clovis-mo** [comment]
Hi, I was looking for a way to achieve this in my migrations.
This is what i came up with

```
Schema::table('orders', function (Blueprint $table) {
    $table->unsignedBigInteger('number')->unique()->nullable()->after('id');
});

Schema::table('orders', function (Blueprint $table) {
    $table->unsignedBigInteger('number')->change()->autoIncrement()->from(5000)->primary(false);
});
```

**Reply by alazark94** [reply]
Did this actually work for you?

**Reply by clovis-mo** [reply]
Yes. It worked for me perfectly

**Comment by pintend** [comment]
Another possible solution (works with singlestore)

```php
Blueprint::macro('autoBigIncrementOnly', function (string $column) {
    /**
     * @var \Illuminate\Database\Schema\Blueprint $this
     */
    return $this->addColumn('autoBigIncrementOnly', $column)->index();
});

SingleStoreSchemaGrammar::macro('typeAutoBigIncrementOnly', function () {
    return 'BIGINT UNSIGNED AUTO_INCREMENT';
});
```

```php
$table->autoBigIncrementOnly('id');
```
